piccolo_api/crud/endpoints.py

```diff
@@ -478,19 +506,101 @@ async def _get_all(
                 },
                 status_code=403,
             )
+
+        # Raise an error if both __page and __cursor are specified
+        if "__cursor" in params and "__page" in params:
+            return JSONResponse(
+                {
+                    "error": (
+                        "You can't have both __page and __cursor as query parameters"
+                    ),
+                },
+                status_code=403,
+            )
+
+        try:
+            # query where limit is equal to page_size plus one
+            query = query.limit(page_size + 1)
+            rows = await query.run()
+            # initial cursor
+            next_cursor = self.encode_cursor(str(rows[-1]["id"]))
+        except IndexError:
+            # for preventing IndexError in piccolo_admin filtering
+            next_cursor = ""
```

Comment:
What do you think about having some way for the user to request cursor pagination?

If we automatically do it, it adds some overhead to each request, and might not be required.

What do you think about only doing it if a cursor is passed in? To initiate it the first time, an empty cursor could be passed in.

Reply:
@dantownsend Like this:
```
if cursor:
    try:
        # query where limit is equal to page_size plus one
        query = query.limit(page_size + 1)
        rows = await query.run()
        # initial cursor
        next_cursor = self.encode_cursor(str(rows[-1]["id"]))
    except IndexError:
        # for preventing IndexError in piccolo_admin filtering
        next_cursor = ""
```
Unfortunately if I do that I'm getting wrong results

Reply:
It would be good to try and debug this - it might be easier when we have unit tests.

We might be able to copy over this one, and modify it:

https://github.com/piccolo-orm/piccolo_api/pull/35/files#diff-9a76db3e8763a9778cd3d34f47be9819e44a3f5e78fa6e80c9d92c692a767b47

It's not the end of the world, but it would be nice to isolate the cursor logic, so it's only run when needed.

Reply:
I've started writing some tests (based on your link) so we'll see what happens.

Reply:
Awesome - if you need any help let me know.

Reply:
I added some tests (latest commit) and they all pass locally. I still don't know how to separate the cursor logic. Maybe you have some ideas about that.

Reply:
Cool, I'll try it out later today.

Reply:
I've played around with it a bit, to try and reduce the amount of cursor logic being run if the cursor isn't required. I've pushed my changes to this pull request.

There are still some more improvements to be made, but it feels a bit cleaner. What do you think?

Reply:
Your changes looks pretty good and a lot cleaner :)